contracts/proxies/CurrencyConvertor.sol

```diff
@@ -1,63 +1,107 @@
 /*
+
     Copyright 2021 dYdX Trading Inc.
+
     Licensed under the Apache License, Version 2.0 (the "License");
     you may not use this file except in compliance with the License.
     You may obtain a copy of the License at
+
     http://www.apache.org/licenses/LICENSE-2.0
+
     Unless required by applicable law or agreed to in writing, software
     distributed under the License is distributed on an "AS IS" BASIS,
     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
     See the License for the specific language governing permissions and
     limitations under the License.
+
 */
 
 pragma solidity ^0.8.0;
-pragma experimental ABIEncoderV2;
-
 
 import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
-import { I_ExchangeWrapper } from "../interfaces/I_ExchangeWrapper.sol";
+import { I_StarkwareContract } from "../interfaces/I_StarkwareContracts.sol";
 
 /**
  * @title CurrencyConvertor
  * @author dYdX
  *
- * Contract for depositing and withdrawing to dYdX L2 in non-USDC tokens.
+ * @notice Contract for depositing to dYdX L2 in non-USDC tokens.
  */
 contract CurrencyConvertor {
-    using SafeERC20 for IERC20;
+  using SafeERC20 for IERC20;
+
+  // ============ State Variables ============
+
+  I_StarkwareContract public immutable STARKWARE_CONTRACT;
+
+  IERC20 immutable USDC_ADDRESS;
+
+  uint256 immutable USDC_ASSET_TYPE;
+
+  // ============ Constructor ============
+
+  constructor(
+    I_StarkwareContract starkwareContractAddress,
+    IERC20 usdcAddress,
+    uint256 usdcAssetType
+  )
+  {
+    STARKWARE_CONTRACT = starkwareContractAddress;
+    USDC_ADDRESS = usdcAddress;
+    USDC_ASSET_TYPE = usdcAssetType;
+
+    // Set the allowance to the highest possible value.
+    usdcAddress.safeApprove(address(starkwareContractAddress), type(uint256).max);
+  }
+
 
   // ============ Events ============
 
-      event LogConvertedDeposit(
-        address indexed account,
-        address source,
-        address exchangeWrapper,
-        address tokenFrom,
-        uint256 tokenFromAmount,
-        uint256 tokenToAmount
-    );
+  event LogConvertedDeposit(
+    address indexed sender,
+    address source,
+    address tokenFrom,
+    uint256 tokenFromAmount,
+    uint256 tokenToAmount
+  );
 
   // ============ State-Changing Functions ============
 
+  /**
+  * Approve an exchange to swap an asset
+  *
+  * @param exchange Address of exchange that will be swapping a token
+  * @param token    Address of token that will be swapped by the exchange
+  */
+  function approveSwap(
+    address exchange,
+    IERC20 token
+  )
+    external
+  {
+    // safeApprove requires unsetting the allowance first.
+    token.safeApprove(exchange, 0);
+    token.safeApprove(exchange, type(uint256).max);
```

Comment:
Regarding this discussion: https://github.com/dydxprotocol/starkex-peripheral/pull/3#discussion_r731133692

You're putting the responsibility of ERC20 approval on the frontend, and it can affect the UI and design considerations if in the edge case the user has to send an extra transaction for approval.

I think a good solution may be to add a function approveSwapAndDeposit() which just calls the two functions together, so that there's no difference from the UI